iib/workers/tasks/build.py

```diff
@@ -337,6 +341,104 @@ def _get_resolved_image(pull_spec):
     return pull_spec_resolved
 
 
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_present_bundles(from_index, base_dir):
+    """
+    Get a list of bundles already present in the index image.
+
+    :param str from_index: index image to inspect
+    :param base_dir: base directory to create temporary files in
+    :return: list of tuples containing bundle names and tags
+    :rtype: list(tuple)
+    :raises IIBError: if any of the commands fail
+    """
+    # Create index image container
+    cidfile = os.path.join(base_dir, 'cidfile.txt')
+    run_cmd(
+        ['podman', 'create', '--cidfile', cidfile, from_index],
+        exc_msg='Failed to create index image container',
+    )
+    with open(cidfile, 'r') as f:
+        cid = f.read()
+    os.remove(cidfile)
+
+    # Get database file from it
+    try:
+        data = json.loads(
+            run_cmd(
+                ['podman', 'image', 'inspect', from_index], exc_msg='Failed to fetch image metadata'
+            )
+        )
+        db_path = data[0]['Config']['Labels']['operators.operatorframework.io.index.database.v1']
+        run_cmd(
+            ['podman', 'cp', f'{cid}:{db_path}', base_dir],
+            exc_msg='Failed to copy DB file from the container',
+        )
+    finally:
+        run_cmd(['podman', 'rm', cid], exc_msg='Failed to delete the container')
+    local_path = os.path.join(base_dir, db_path.split('/')[-1])
+
+    free_port = get_free_port(50051, 50051 + 1000)
+    if not free_port:
+        error_msg = 'No free port is available'
+        log.error(error_msg)
+        raise RuntimeError(error_msg)
+
+    # Get present bundle information
+    cmd = ['opm', 'registry', 'serve', '-p', str(free_port), '-d', local_path]
+    log.debug('Running the command "%s"', ' '.join(cmd))
+    rpc_proc = subprocess.Popen(cmd)
+    MAX_TRIES = 5
+    for i in range(MAX_TRIES):
+        try:
+            time.sleep(1)
+            bundles = run_cmd(
+                ['grpcurl', '-plaintext', f'localhost:{free_port}', 'api.Registry/ListBundles'],
+                exc_msg='Failed to get bundle data from index image',
+            )
+            break
+        except IIBError:
+            if i == MAX_TRIES - 1:
+                log.error('Data cannot be retrieved from the index image database gRPC.')
+                raise
+        finally:
+            rpc_proc.kill()
+
+    # Transform returned data to parsable json
+    present_bundles = json.loads('[' + bundles.replace('\n}\n{\n', '\n},\n{\n') + ']')
+    present_bundles = [(x['packageName'], x['version']) for x in present_bundles]
+    return present_bundles
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _filter_bundles(present_bundles, bundles):
```

Comment:
Can we please rename this to `get_missing_bundles` instead? `_filter_bundles` seems a bit generic

Reply:
Or `filter_missing bundles`?

Reply:
done